packages/editor-ui/src/views/NodeView.vue

```diff
@@ -3243,12 +3327,51 @@ export default mixins(
 				this.$telemetry.trackNodesPanel('nodeView.createNodeActiveChanged', { source, createNodeActive, workflow_id: this.workflowsStore.workflowId });
 			},
 			onAddNode({ nodeTypeName, position }: { nodeTypeName: string; position?: [number, number] }) {
-				this.addNode(nodeTypeName, { position });
+				this.addNode(nodeTypeName, { position }, true);
 			},
 			async saveCurrentWorkflowExternal(callback: () => void) {
 				await this.saveCurrentWorkflow();
 				callback?.();
 			},
+			setSuspendRecordingDetachedConnections(suspend: boolean) {
+				this.suspendRecordingDetachedConnections = suspend;
+			},
+			onMoveNode({nodeName, position}: { nodeName: string, position: XYPosition }): void {
+				this.workflowsStore.updateNodeProperties({ name: nodeName, properties: { position }});
+				setTimeout(() => {
+					const node = this.workflowsStore.getNodeByName(nodeName);
+					if (node) {
+						this.instance.repaintEverything();
+						this.onNodeMoved(node);
+					}
+				}, 0);
+			},
+			onRevertAddNode({node}: {node: INodeUi}): void {
+				this.removeNode(node.name, false);
+			},
+			async onRevertRemoveNode({node}: {node: INodeUi}): Promise<void> {
+				const prevNode = this.workflowsStore.workflow.nodes.find(n => n.id === node.id);
+				if (prevNode) {
+					return;
+				}
+				// For some reason, returning node to canvas with old id
+				// makes it's endpoint to render at wrong position
+				node.id = uuid();
+				await this.addNodes([node]);
+			},
+			onRevertAddConnection({ connection }: { connection: [IConnection, IConnection]}) {
+				this.suspendRecordingDetachedConnections = true;
+				this.__removeConnection(connection, true);
+				this.suspendRecordingDetachedConnections = false;
+			},
+			async onRevertRemoveConnection({ connection }: { connection: [IConnection, IConnection]}) {
+				this.suspendRecordingDetachedConnections = true;
+				this.__addConnection(connection, true);
```

Comment:
what if you tracked history changes only in `__addConnection` instead of `connection bind` event? then you could pass `trackHistory: false` and not track events.. and not need this flag to suspend recoding events.. same question applies to other events.. 
not a huge fan of this flag because it seems flaky and easy thing to mess up...

Reply:
I don't like it ether, but found out not all connections are added and removed ind `__addConnection` and `removeConnection` methods, some are added and removed in those events. For example, if you drag a connection on an empty canvas space, it is removed in the `connectionDetached` event (among other situations, hence the flag), also when reconnecting nodes, some connections are pushed straight to store in `conncection` event. This made implementation rely on those flags which is a bit of chasing my own tail situation but I couldn't figure out better way to cover all cases.